src/clients/python/experimental_api_v2/examples/v2_image_client.py

```diff
@@ -373,20 +407,73 @@ def requestGenerator(supports_batching, input_name, output_name, c, h, w, format
     # start over with the first images until the batch is filled.
     requests = []
     responses = []
+    user_data = UserData()
+
+    # Used to hold the greenlets running the intference in HTTP
+    # async operation.
+    greenlets = []
+
+    sent_count = 0
 
     # Send request
     try:
+        if FLAGS.streaming:
+            triton_client.start_stream(partial(completion_callback, user_data))
+
         for inputs, outputs, model_name, model_version in requestGenerator(
-                max_batch_size > 0, input_name, output_name, c, h, w, format, dtype, FLAGS):
-            responses.append(
-                triton_client.infer(FLAGS.model_name,
-                                    inputs,
-                                    model_version=FLAGS.model_version,
-                                    outputs=outputs))
+                max_batch_size > 0, input_name, output_name, c, h, w, format,
+                dtype, FLAGS):
+            sent_count += 1
+            if FLAGS.streaming:
+                triton_client.async_stream_infer(
+                    FLAGS.model_name,
+                    inputs,
+                    model_version=FLAGS.model_version,
+                    outputs=outputs)
+            elif FLAGS.async_set:
+                if FLAGS.protocol.lower() == "grpc":
+                    triton_client.async_infer(FLAGS.model_name,
+                                              inputs,
+                                              partial(completion_callback,
+                                                      user_data),
+                                              model_version=FLAGS.model_version,
+                                              outputs=outputs)
+                else:
+                    greenlets.append(
```

Comment:
Ideally we should not expose gevent/greenlet through our API since it ties us to that implementation. Instead we could wrap the greenlet in our own class. But why return the greenlet? We already have a callback mechanism to indicate completion. Can we just bind the callback to the greenlet completion (there appears to be an API for that)?

Reply:
We were calling the gevent version of sleep to relinquish control to underlying greenlets. These work with cooperative multi-tasking. We do have attached callback. But the callback will not be invoked if the current greenlet(the main execution stream) does not relinquish control. We were doing that with gevent.sleep(). This kind of exposes the underlying greenlet implementation but will lead to continuous polling. The join() call on the underlyinh greenlet is less cpu intensive. I can fallback to our previous gevent.sleep() implementation. The thing is the greenlet will not execute until and unless someone releases the control to it.